[Add] '중간미션' 의 '자판기 애플리케이션' 코드를 추가한다

antaehyeon opened the pull request:
코드 설계는 README 에 포함되어 있습니다.
블럭박스(체크)로 작성된 부분이 코드를 작성하면서 추가된 부분입니다.

antaehyeon committed:
[Add] '중간미션' 의 '자판기 애플리케이션' 코드를 추가한다

코드 설계는 README 에 포함되어 있습니다.

crongro commented on README.md:
```diff
@@ -1,2 +1,49 @@
+
+  - #### 설계
+
+    - 함수부터 무엇이 필요할지
```
설계는 무엇이 필요한지는 잘 적었는데요.

어떻게 동작하는지에 대한 동작과정(절차)는 잘 안보이네요.
다음엔 미리 이런 부분도 한번 적은 후에 코딩을 해보세요.

antaehyeon replied:
확인하였습니다 👍

crongro commented on vendingMachine.js:
```diff
@@ -0,0 +1,86 @@
+
+function insertCoin(money) {
+  currentMoney += money;
+  console.log(" " + money + " 원이 투입되었습니다.");
```
반복적이니, log함수를 만들어서 log("") 이렇게 쓰시면 좋겠군요.

antaehyeon replied:
확인하였습니다 👍

crongro commented on vendingMachine.js:
```diff
@@ -0,0 +1,86 @@
+    return;
+  }
+
+  for (var index in drinks) {
```
배열은 for-in을 잘 안써요.(이유가 있는데 참고로 한번 찾아보세요. 반복과정에서 불필요한 것이 나올 확률도 아주조금이지만 있어요.)
for 나 forEach와 같은 메서드를 씁니다.

antaehyeon replied:
검색 후, README에 추가하였습니다 👍

crongro commented on vendingMachine.js:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  for (var index in drinks) {
+    var _name = drinks[index].name;
```
_는 private인 경우인가요? 대부분이 이런 경우이면 _ 를 안붙이는 것도 방법이겠어요. 코드가 보기 좋진 않네요.

antaehyeon replied:
확인하였습니다 👍

antaehyeon replied:
제가 _ 를 사용한 주된 이유는 다음과 같습니다.
1. 네이밍이 중복되어서 구별을 주기위해서 사용했습니다.

혹시, 전역변수와 네이밍이 겹칠때, 지역변수는 어떤식으로 네이밍을 해야하는지 알려주시면 감사하겠습니다 !!

crongro commented on vendingMachine.js:
```diff
@@ -0,0 +1,86 @@
+
+  console.log(" " + drinkName + "(이/가) 나왔습니다.");
+  console.log(" 현재 잔돈 : " + currentMoney);
+  newLine();
```
newLine이 필요한 경우에는 console.log안에서 처리하는 게 더 좋을 듯.

antaehyeon replied:
확인하였습니다 👍

crongro commented on vendingMachine.js:
```diff
@@ -0,0 +1,86 @@
+
+  console.log("사용 가능한 음료수 목록 => ");
+
+  if (currentMoney < 500) {
```
500을 함수안에서 빼는 건 가능할까요? 
바뀔수도 있는 숫자라..

antaehyeon replied:
아, 그 부분을 고려하지 못했군요.
수정해서 재 Push 해보도록 하겠습니다 👍

crongro commented on vendingMachine.js:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    if (_price <= currentMoney) {
+      var _price = (!_price) ? _price = "재고없음" : _price;
```
_price = "재고없음" 
뭔가 이상한데요, 중간에 이코드가 필요한가요?

antaehyeon replied:
저의 음료수 출력함수는 printAvailableDrinkList () 입니다.
형식은 음료수이름("가격") 으로 출력되는데,

1. 재고없음을 판단하기 위해서 일단 음료수의 수량을 체크합니다.
2. 이후 해당 음료수의 가격을 0으로 맞춰버리고
3. 출력할 때, 만약 가격이 0인 경우, 가격대신 "재고없음" 이라는 문구가 대신 들어가도록 코딩하였습니다.

crongro replied:
아 네 이해했고요.
코드를 좀더 줄일 수 있겠네요.

var _price = (!_price) ? "재고없음" : _price;

crongro commented on vendingMachine.js:
```diff
@@ -0,0 +1,86 @@
+    var _drinkName = drinks[index].name;
+
+    if (drinkName === _drinkName) {
+      var _price = drinks[index].price;
```
var로 선어한 경우, 함수 맨 위로 호이스팅이 발생합니다.
조건문 안에 선언하지 말고, 함수바로 아래 변수를 모두 선언해두는 게 더 좋아요.

function a() {
 price = 0;
 num = 0;
....
}

antaehyeon replied:
호이스팅!
실수를 했네요. 다음부터는 꼭 확인하면서 코드를 작성하겠습니다 👍

crongro reviewed (changes requested):
금방 구현하셨네요~ 
수정해보세요.

antaehyeon committed:
[Fix] '자판기 애플리케이션' 코드를 피드백에 반영하여 수정한다

crongro commented on vendingMachine.js:
```diff
@@ -13,74 +13,91 @@ var currentMoney = 0;
-    var _name = drinks[index].name;
-    var _price = drinks[index].price;
-    var _num = drinks[index].num;
+  drinks.forEach(function (element, index, array) {
```
이코드가 반복을 돌면서 계속 name,price, num에 변경된 값을 업데이트 해주는건데 맞나요? 결국 마지막값이 할당되는거 같네요.

antaehyeon replied:
네 맞습니다. 결국에는 마지막 값이 할당됩니다.

해당 코드의 목적은 사용 가능한 음료수 목록을 출력하는 코드입니다.
그래서 배열을 forEach 문으로 하나씩 탐색하면서, 출력하고
다음 음료의 배열을 탐색하면서 출력하고

그래서 모든 음료수와 가격이 나옵니다 :)

crongro commented on vendingMachine.js:
```diff
@@ -13,74 +13,91 @@ var currentMoney = 0;
+  var drinkPrice;
+
+  drinks.forEach(function (element, index, array) {
+    drinkPrice = array[index].price;
```
이것도 마찬가지인데, drinks를 돌면서 마지막 값이 drinkPrice에 반영되는 코드네요. 의도한게 이상하네요.

antaehyeon replied:
아 맞습니다.
그 저번 코드의 피드백 중 

500을 함수안에서 빼는 건 가능할까요?
바뀔수도 있는 숫자라..

이 부분을 고려하기 위해서 코드를 작성했습니다.
배열에 어떤 음료수가 추가되어도 가장 가격이 낮은 음료수의 값을 찾기 위해서
findLowestPrice() 라는 메서드를 만들게 되었구요.

init 값을 10000 으로 지정한 후, forEach문으로 음료수의 배열을 탐색하면서
가격이 낮은 것이 계속 값으로 들어가도록 하여, 가장 낮은 음료수의 값을 찾고

위에서 가장 낮은 음료수의 가격보다 현재 잔돈이 없을경우를 체크하기 위해서
이렇게 코드를 작성하게 되었습니다.

crongro replied:
findLowestPrice에서 최소값을 찾는 로직이 이상하네요.
findLowestPrice를 직접 호출테스트 하면서 잘 동작하는지 확인해보세요
그리고 지금 findLowestPrice 은 반환하는 것도 없네요.

crongro commented on vendingMachine.js:
```diff
@@ -13,74 +13,91 @@ var currentMoney = 0;
+
+  drinks.forEach(function (element, index, array) {
+    drinkPrice = array[index].price;
+    lowestPrice = (drinkPrice < lowestPrice) ? lowestPrice = drinkPrice : lowestPrice = lowestPrice;
```
삼항연산자 중간에 lowestPrice은 생략가능해요.

antaehyeon replied:
삼항연산자에 대해서 제대로 알고갑니다.
감사합니다 👍

crongro reviewed (changes requested):
수정해보세요~
readme 정리를 project마다 하고 계신데, 
본인이 정리하는데 이게 도움이 되셔서 하는것이죠?

antaehyeon commented:
아, 저의 습관이긴 합니다.
동영상 강의가 있으면, 동영상 강의를 듣기전에 미리한번 내용을 타이핑하면서 생각해봅니다.
그리고 동영상 강의때는 동영상강의만 듣는식으로 진행하고 있습니다 :)
(동영상 강의를 들으면서 정리하니까, 계속 멈췄다가 타이핑하니 흐름이 끊겨서 집중이 잘 안되더라구요.)

그러면, 동영상 강의를 들을 때 정리했던 것들이 한번 더 새겨지면서 기억에 오래 남아서 그런것 같습니다.

나중에 찾아보기도 편하구요 👍

antaehyeon committed:
[Fix] '자판기 애플리케이션' 코드를 피드백에 반영하여 수정한다

삼항연산자 수정 및 변수 네이밍 수정

antaehyeon committed:
[Fix] '자판기 애플리케이션' 코드중 음료의 최소가격을 찾는 함수에 대한 피드백을 반영한다

최소 가격을 찾는 함수에 대해 로그를 찍어서 확인해봤습니다.
https://imgur.com/NpiSZHb

antaehyeon commented:
음.. 로직에 대한 테스트를 진행해봤습니다.
https://imgur.com/NpiSZHb

분명 코드 제출하기전에 항상 결과값을 확인하는데 정상적으로 출력됬거든요.
근데 크롱 피드백을 받고나서 함수를 확인했는데 return 값이 없는데
왜 디버그 콘솔과 출력창에 정상적으로 출력되는지 모르겠네요..
당황스럽네요..
https://imgur.com/X6UmD9I

antaehyeon committed:
[Add] '중간 미션' 중 '로또 생성기' 에 대한 설계를 추가한다

antaehyeon committed:
[Add] '중간 미션' 중 '로또 생성기' 에 대한 1차완성 코드를 추가한다

antaehyeon committed:
[Add] '중간 미션' 중 '로또 생성기' 에 대한 정렬코드를 추가한다.

나중에 알고리즘 공부할 때, 내장함수 대신 직접 함수로 구현해볼 것

antaehyeon committed:
[Add] '중간 미션' 중 '학습 키워드'에 대해 아는대로 작성한다

'중간미션' 학습키워드에 대한 내용을 아는대로 적어봤습니다.
이후, 검색을 통해서 정확한 내용을 짚어갈 것 입니다.

crongro commented on README.md:
```diff
@@ -1,2 +1,253 @@
+
+  - **!! 은 무엇을 의미하는가? 어떻게 활용할 수 있을까?**
+
+    - 음.. 이건 모르겠네요
```
실험을 해보세요~ 
!!true 
!!3
등등

antaehyeon replied:
확인하였습니다 👍

crongro commented on README.md:
```diff
@@ -1,2 +1,253 @@
+
+  - **==와 ===의 차이는 정확히 무엇인가?**
+
+    - `==` 는 타입상관 없이 값만 비교한다 (?)
```
차이점을 좀더 공부해보시면 좋겠네요.

antaehyeon replied:
확인하였습니다 👍

crongro commented on lotto.js:
```diff
@@ -0,0 +1,156 @@
+  // switch-case 문을 이용해서
+  // 아까 setLuckyNumber 에서 
+
+  var yield;
```
yield는 예약어같은 느낌이네요. 쓰지 않도록하는게 좋겠어요.

crongro commented on lotto.js:
```diff
@@ -0,0 +1,156 @@
+
+function calRateOfReturnToInvestment(three, four, five, six) {
+  var money = lottos.length * 1000;
+  var totalPrizeMoney = (three * 5000) + (four * 50000) +
```
three 는 5000원이고, four은 50000이고,, 이런 데이터관계를 객체로 만들어서 코드에서 사용해보세요.
switch문과 이런코드가 없어질 수 있을 겁니다.

antaehyeon replied:
확인하였습니다 👍
다음.. 커밋때 확인 부탁드리겠습니다

crongro commented on vendingMachine.js:
```diff
@@ -0,0 +1,118 @@
+
+function selectItem(drinkName) {
+
+  var name;
```
초기값을 셋팅해두면 어떤 타입인지 좀더 알기 편해요.

antaehyeon replied:
확인하였습니다 👍

crongro commented on vendingMachine.js:
```diff
@@ -0,0 +1,118 @@
+    return;
+  }
+
+  drinks.forEach(function (element, index, array) {
```
drinks가 전역공간에 있긴하지만, 
이 함수의 인자로 받아서 동작되게 하면 좋겠네요.

antaehyeon replied:
확인하였습니다 👍
다음 Push 때, 이를 반영하였으니 맞게 구현했는지 확인해주시면 감사하겠습니다!

antaehyeon replied:
질문이 있습니다.
똑같은 코드인데 함수의 인자로 Object 를 잘 전달받다가, 메서드를 한번 더 실행하면
해당 Object가 undefined 가 되면서, 코드수행이 멈춰버립니다.

제가 놓치고 있는것이 있나요?

antaehyeon committed:
[Add] '중간 미션' 중 '학습 키워드'에 대해 검색을 토대로 작성한다

'중간미션' 학습키워드에 대한 내용을 검색하면서 작성했습니다.

crongro commented on lotto.js:
```diff
@@ -0,0 +1,156 @@
+    log(lotto);
+  }
+
+  winCountCheckArray = new Array(0, 0, 0, 0, 0, 0, 0,
```
여기나열된 배열은 무엇일까요??? 
이걸 제거하고 구현할 수 있으면 좋겠어요.

antaehyeon replied:
일단, winCountCheckArray는 로또가 당첨된 갯수를 체크하는 배열입니다.
선언과 동시에 0으로 초기화 하려는 목적이였습니다.
이를 반영한 코드를 Push 하도록 하겠습니다 👍

crongro commented on lotto.js:
```diff
@@ -0,0 +1,156 @@
+    if (lotto.indexOf(randNum) === -1) {
+      lotto.push(randNum);
+    } else {
+      continue;
```
continue 가 필요 없을 듯.

antaehyeon replied:
확인하였습니다 👍

crongro commented on lotto.js:
```diff
@@ -0,0 +1,156 @@
+  printWinningStatistics(winCountCheckArray);
+}
+
+function findWinNumber(num) {
```
find가 아니고 무언가 증가시켜주는 거 같은데요.
그럼 함수이름도 그거에 맞게 변경~

antaehyeon replied:
확인하였습니다 👍

crongro commented on lotto.js:
```diff
@@ -0,0 +1,156 @@
+  
+}
+
+function matchWinNubmer(array) {
```
함수이름을 좀더 구체적인 의도로 표현. 
match~~ 는 두리뭉실해요.

antaehyeon replied:
확인하였습니다 👍

crongro reviewed (changes requested):
중간리뷰~추가

antaehyeon committed:
[Add] 'scope와 closure 기초' 중 첫번 째 make 함수를 작성한다

antaehyeon committed:
[Add] 'scope와 closure 기초' 중 두번 째 make 함수를 작성한다

rect 메서드와 circle 메서드가 포함되어 있는 make 함수를 작성했습니다.

antaehyeon committed:
[Fix] 중간미션에 대한 피드백을 반영한다

음료자판기와 로또 생성기에 대한 피드백을 반영합니다.
해당 피드백의 내용은 PR의 링크로 대체합니다.
https://github.com/code-squad/javascript-grade/pull/11

crongro reviewed (changes requested):
이거 리뷰가 남은 걸 제가 놓친거 같네요.

var 대신에 let, const 를 사용해보고요
lotto의 switch 문을 없애보세요.

antaehyeon committed:
refactor : 피드백 반영

var 대신에 let, const 사용 (let, const, var의 차이 확인)
lotto의 switch 문 대체

crongro reviewed (approved):
좋아요.

const/let 을 앞으로는 사용하시고요.var와의 차이점은 잘 알고 계시고요. const를 먼저 쓰고 재할당이 필요한 경우 let을 보통 씁니다.

주석은 지금코드보다 적게 쓰세요. 코드는 코드로 표현되는게 가장 좋아요.
